[wptrunner] Reject tests that almost time out in stability check

zcorpan opened the pull request:
Fixes #9972.

~TODO: allow longer time for timeout=long tests.~

zcorpan committed:
[wptrunner] Reject tests that almost time out

In stability check (wpt run --verify), reject tests that almost time
out, i.e. take more than 80% of the timeout allowed to run). These tests
will be listed in a new section, "slow tests", in the output.

Fixes #9972.

zcorpan commented:
I was hoping @gsnedders or someone could help me figure out how to check for `timeout=long`

zcorpan commented:
Let's use 30s for now.

gsnedders commented:
> I was hoping @gsnedders or someone could help me figure out how to check for timeout=long

I'm not sure there's any easy way to get at that from here, because if I'm not mistaken we've not got as far as loading the manifest at the point of running the stability entry point.

Hexcles commented:
We almost want to count the longest duration for tests with normal timeout and tests with long timeout separately (instead of using different threshold depending on what the slowest test is), because otherwise the longest running test is very likely one with long timeout and we miss those tests with normal timeout but perhaps closer to timeout.

As for the implementation, @gsnedders is right; it's very hard to consult the manifest here. I'm thinking perhaps the wptrunner can add the timeout of the test as an additional field to its `test_end` messages. This field should be ignored by the other log handlers (to avoid bloating the log or report), but can be used by the stability check.

zcorpan commented:
>  I'm thinking perhaps the wptrunner can add the timeout of the test as an additional field to its `test_end` messages. 

I've tried to poke around a bit, but can't figure out how to do this. Any pointers?

zcorpan commented:
Thanks, I think I got it working (but doesn't do anything with `timeout_multiplier` unless that's already baked in).

Hexcles reviewed (approved):
Cool! I'm giving my r+, but please also wait for opinions from others.

zcorpan commented:
@Hexcles can you merge this if there has been no review by June 29? Also, if there *are* review comments, can you address them, please?

gsnedders commented:
I'd like to see what @jgraham has to say, but he's on holiday till after then IIRC.

zcorpan commented:
Ok to wait for James.

Hexcles commented:
@zcorpan alright, so let's wait for more reviews. If it's something easy to fix, I'll help you address the comments and land the PR when you're away. Otherwise, I'll wait till you're back.

Meanwhile, could you take a look at the high-level discussions in #9972 first?

jgraham commented:
I haven't reviewed this because I'm away, but as noted on the original issue, I'm pretty skeptical of failing travis because of long-running tests. In particular it's hostile to tests for features that aren't implemented in any browser we run in CI which are more likely to time out.

I think we should discuss some mechanism to warn for this kind of thing rather than hard-failing.

ricea commented:
Indeed, "Failing by timeout" is sometimes the most natural way to write a test (for example, a fetch abort test). I was assuming that reviewers would be able to apply their judgement in these cases.

zcorpan commented:
Hmm I didn't intend do fail the check for TIMEOUT.

Hexcles commented:
So to clarify (based on my understanding):

* Tests that consistently time out won't fail the check.
* Tests that time out sometimes will fail the check (which is already the case).
* (NEW) Tests that never timeout (during the 10 reruns) but are very close to timeout will now fail as they are likely to be flaky.

And I think this is a reasonable approach to prevent flakiness with minimal cost of productivity.

foolip commented:
@jugglinmike, while @zcorpan is away, can you put this on the Bocoup list of things to potentially prioritize?

Hexcles commented:
@foolip I reviewed the code in detail and am reasonably familiar with related code. I agreed to help @zcorpan to rebase, fix nits and land this PR when he's away. I think this is now still awaiting more review. Ping @gsnedders @jgraham

jugglinmike commented:
Very good. I'll let you take it from here :)

zcorpan commented:
I think this has two outstanding todos.

* move logic to test_start
* don't fail for consistent TIMEOUT.

jgraham commented on tools/wptrunner/wptrunner/testrunner.py:
```diff
@@ -582,6 +582,10 @@ def test_ended(self, test, results):
+        if not file_result.extra:
+            file_result.extra = { "test_timeout": test.timeout }
+        else:
+            file_result.extra.test_timeout = test.timeout
```
Does this work? I would expect it to be `file_result.extra["test_timeout"]` unless `extra` isn't a dictionary for some reason.

Hexcles committed:
Take timeout_multiplier into account

Hexcles committed:
Longest duration for each status of each test

Previously we record the longest duration of all tests. However, each
test can have different timeout. Besides, test runs with different
results should be considered differently; runs that time out should be
ignored.

Therefore, we now keep track of the longest duration for each result
status of each test.

Also add a unit test.

jgraham commented on tools/wptrunner/wptrunner/stability.py:
```diff
@@ -91,9 +106,18 @@ def is_inconsistent(results_dict, iterations):
+def is_slow(test):
+    # We are interested in tests that almost time out (i.e. likely to be
+    # flaky). Tests that consistently time out are not considered "slow".
+    if test["status"].keys() == ['TIMEOUT']:
```
I think `CRASH` should also be excluded because we can change the status from `TIMEOUT` to `CRASH` if we detect evidence of a crash after the test finished. In general I wonder if testing 

Also, it seems arguable that the handling of a test that sometimes times out and sometimes doesn't is wrong; if we get `PASS`, `PASS`, `PASS`, `TIMEOUT`, `PASS` then we'll call the test slow. But if the passes take 1s and the timeout is 10s then we are probably seeing a race condition that causes the test to hang sometimes, and the obvious solution to a slow test (Add a longer timeout) isn't going to help. So I think we should either store the duration per result and filter here, or only update the duration when we see a non-TIMEOUT, non-CRASH result. The latter is easier, but the former could be used to provide more detailed logging if we want.

Hexcles replied:
You're right about `CRASH`.

As for the second example (inconsistent results), it won't actually happen as we call `is_inconsistent` before `is_slow`, so such tests will be considered as flaky instead of slow. But I do see the code can be confusing and error-prone as the prerequisite isn't obvious. I'll tweak it a bit.

Hexcles committed:
Fix ci/check_stability.py

Hexcles commented:
@jgraham I think I addressed your comment this time. PTAL at `find_slow_status` (the new version of `is_slow`) in particular.

And you can see stability checks rejecting the example slow test in action:
https://travis-ci.org/web-platform-tests/wpt/jobs/422774773#L764
https://travis-ci.org/web-platform-tests/wpt/jobs/422774775#L686

jgraham commented on tools/wptrunner/wptrunner/stability.py:
```diff
@@ -91,9 +107,29 @@ def is_inconsistent(results_dict, iterations):
+        runs almost time out.
+    """
+    threshold = test["timeout"] * FLAKY_THRESHOLD
+    for status in ['PASS', 'FAIL', 'OK', 'ERROR', 'INTERNAL-ERROR']:
```
I don't think we can actually get `INTERNAL-ERROR` as a status here; it's alwasy converted to `ERROR`.

jgraham commented on tools/wptrunner/wptrunner/stability.py:
```diff
@@ -91,9 +107,29 @@ def is_inconsistent(results_dict, iterations):
+    threshold = test["timeout"] * FLAKY_THRESHOLD
+    for status in ['PASS', 'FAIL', 'OK', 'ERROR', 'INTERNAL-ERROR']:
+        if (status in test["longest_duration"] and
+                test["longest_duration"][status] > threshold):
```
Weird indentation (yeah I know pyflakes likes it, but I think it's a bad rule which we don't enforce).
